tokio-util/src/sync/cancellation_token.rs

```diff
@@ -77,15 +77,62 @@ pin_project! {
     /// [`CancellationToken`] by value instead of using a reference.
     #[must_use = "futures do nothing unless polled"]
     pub struct WaitForCancellationFutureOwned {
+        // `Notified` references to the cancellation_token internally,
+        // but camouflages the relationship with `'static`.
+        // This can lead to Undefined Behavior, breaking the pointer aliasing rules
+        // if the reference is dangling.
+        // This is fixed here by the use of `MaybeDanglingFuture`, which prevent
+        // the compiler to generate the UB.
         // Since `future` is the first field, it is dropped before the
         // cancellation_token field. This ensures that the reference inside the
-        // `Notified` remains valid.
+        // `Notified` remains valid, hance the safety requirements of
+        // `MaybeDanglingFuture` are satisfied.
         #[pin]
-        future: tokio::sync::futures::Notified<'static>,
+        future: MaybeDanglingFuture<tokio::sync::futures::Notified<'static>>,
         cancellation_token: CancellationToken,
     }
 }
 
+/// A wrapper type that avoids implicitly asserting the contents are currently valid.
+/// It's safe to contain a dangling reference, as long as it is not used.
+///
+/// # Safety
+///
+/// The inner type must be dropped before the data it references to.
+///
+// Reference
+// See <https://users.rust-lang.org/t/unsafe-code-review-semi-owning-weak-rwlock-t-guard/95706>
+//
+// TODO: replace this with an official solution once RFC #3336 or similar is available.
+// <https://github.com/rust-lang/rfcs/pull/3336>
+#[repr(transparent)]
+struct MaybeDanglingFuture<T>(core::mem::MaybeUninit<T>);
```

Comment:
I would move this into another file.

Reply:
`MaybeDangling` was renamed and moved to a new file under `cancellation_token/` - I'm not sure if this is the best place, let me know if you think there's a better one :)

Reply:
One convention we use in the main Tokio crate is to have a `src/util` folder for this kind of thing. I guess the same could make sense here. (We already have a `mod util {}` in `lib.rs`, so it would have two things if we moved it there.)

Reply:
is it ok for you if I add it to the mod here and open another PR to extract the mod to a folder?

Reply:
Either is fine.

Reply:
I decided to go with the `util` folder because it fits best with the organization/test